src/lib.rs

```diff
@@ -253,35 +253,39 @@ fn expand_to_file(
         .truncate(true)
         .open(dest.as_path())?;
 
-    let Ok(mut f) = file_guard::try_lock(f.file_mut(), file_guard::Lock::Exclusive, 0, 64) else {
-        // the digest of the file will not match if the content to be written differed, hence any existing lock
-        // means we are already writing the same content to the file
-        if verbose {
-            eprintln!("expander: already in progress of writing identical content to {} by a different crate", dest.display());
-        }
-        // now actually wait until the write is complete
-        let _lock = file_guard::lock(f.file_mut(), file_guard::Lock::Exclusive, 0, 64)
-            .expect("File Lock never fails us. qed");
+    // Force implicit drop of FileGuard before rustfmt to prevent intermittent partial file locking issues on Windows (os error 33)
```

Comment:
Can you elaborate what exactly is locking up and why?

Reply:
The FileGuard maintains an exclusive lock on the first 64 bytes of the file for the digest.
On Windows, if this lock is still held when rustfmt tries to access the file, it causes "process cannot access the file" errors (os error 33). This occurs because rustfmt needs full access to read and modify the file, but can't while we hold the lock. The lock needs to be explicitly released before calling external processes that need to access the file.

Reply:
Re-reading the documentation of `file-guard`, I think the implementation above is flawed, although I don't quite understand why it exhibits the deadlock only under windows.

Reply:
Ok, I think I've found the explanation in the `file-guard` documentation:

"A cross-platform library for simple **advisory** file locking.
...
Anything that can **Deref or DerefMut to a File** can be used with the FileGuard (i.e. Rc<File>):"

(Emphasis mine.)

Per Wikipedia:
"For applications that use the __file read/write [APIs]__(https://en.wikipedia.org/wiki/API) __in Windows, byte-range locks are enforced__ (also referred to as __mandatory locks__) by the file systems that execute within Windows. For applications that use the file mapping [APIs](https://en.wikipedia.org/wiki/API) in Windows, byte-range locks are not enforced (also referred to as ___advisory locks___.)
...
Although some types of locks can be configured to be mandatory, file locks under __Unix__ are by default ___advisory___. This means that cooperating processes may use locks to coordinate access to a file among themselves, but __uncooperative processes are also free to ignore locks and access the file in any way they choose__. In other words, file locks lock out other file lockers only, not I/O."

(https://en.wikipedia.org/wiki/File_locking: Italics in original; bolding is mine.)

This explains why the `file_guard` locking in Windows is mandatory and therefore affects rustfmt, whereas Linux and macOS, being Unix-like, are doing advisory locking, which rustfmt is free to ignore.

I also now realise that dropping the FileGuard before calling rustfmt is not the answer, as it leaves the file unprotected from this point, allowing a second process to write to the file while rustfmt is still formatting it. And yet we need an answer for WIndows.

__I have come up with an alternative solution__ that works in testing and overcomes these obstacles. Instead of using rustfmt on disk, we simply invoke it on stdin and emit the output to stdout.  I've double-checked on the command line that this works (`cat myfile.rs | rustfmt --emit stdout --`).

Only having formatted the content (or not, as requested) do we write it out to the guarded file. For simplicity I write out any comment first, on the reasonable assumption that it does not need to be formatted. Otherwise we will need to concatenate the comment and the content into a single Vec before the formatting step.

``` rust
    // Unchanged up to here

    if verbose {
        eprintln!("expander: writing {}", dest.display());
    }

    // Write out the comment before any formatting to keep things simple,
    //  assuming no need to format a comment.
    if let Some(comment) = comment.into() {
        f.write_all(&mut comment.as_bytes())?;
    }

    // Format content in memory if requested
    let content = if let RustFmt::Yes {
        channel,
        edition,
        allow_failure,
    } = rustfmt
    {
        &format_content(&bytes, channel, edition, allow_failure)?
    } else {
        bytes
    };

    // Now write the content while holding the guard
    f.write_all(&content)?;

    let dest = dest.display().to_string();
    Ok(quote! {
        include!( #dest );
    })
}

use std::process::Stdio;

fn format_content(
    content: &[u8],
    channel: Channel,
    edition: Edition,
    allow_failure: bool,
) -> Result<Vec<u8>, std::io::Error> {
    let mut process = std::process::Command::new("rustfmt");
    if Channel::Default != channel {
        process.arg(channel.to_string());
    }

    let mut child = process
        .arg(format!("--edition={}", edition))
        .arg("--emit=stdout")
        .arg("--") // Signal to read from stdin
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    // Write content to rustfmt's stdin
    if let Some(ref mut stdin) = child.stdin {
        stdin.write_all(content)?;
        // Dropping stdin here signals EOF to rustfmt
    }

    let output = child.wait_with_output()?;
    if !output.status.success() {
        let error = std::io::Error::new(
            std::io::ErrorKind::Other,
            format!(
                "rustfmt failed with exit code {}\nstderr: {}",
                output.status.code().unwrap_or(-1),
                String::from_utf8_lossy(&output.stderr)
            ),
        );
        if allow_failure {
            eprintln!("expander: {}", error);
            Ok(content.to_vec())
        } else {
            Err(error)
        }
    } else {
        Ok(output.stdout)
    }
}
```

Let me know if you want to proceed on this basis.

And Happy New Year :)